SCons/Environment.py

```diff
@@ -2585,51 +2645,94 @@ def __setattr__(self, name, value) -> None:
         setattr(self.__dict__['__subject'], name, value)
 
     # Methods that make this class act like a dictionary.
+
     def __getitem__(self, key):
+        """Return the visible value of *key*.
+
+        Backfills from the subject env if *key* doesn't have an entry in
+        the override, and is not explicity deleted.
+        """
         try:
             return self.__dict__['overrides'][key]
         except KeyError:
+            if key in self.__dict__['__deleted']:
+                raise
             return self.__dict__['__subject'].__getitem__(key)
 
-    def __setitem__(self, key, value):
+    def __setitem__(self, key, value) -> None:
         # This doesn't have the same performance equation as a "real"
```

Comment:
I think this should do the following:
1. if the envvar is in the overrides, it should overwrite that.
2. if the envvar is in the __deleted, it should remove from __deleted and write to overrides
3. otherwise it should write to the parent Environment.

OverrideEnvironment() is not meant to shield the original env, but rather to allow light weight specialization.

Reply:
1. does that now, and in the PR also
2. current and PR both write to `overrides` in this case; PR adds the `__deleted`, and it does clear from there on assignment.
3. current code doesn't do that, nor does the PR.  However, modifying (rather than assigning to) a value does affect the parent environment - that behavior does not differ between current and PR.

So point 3 would change existing behavior somewhat - it says items can only be added to the override dict on creation -- or by modifying the override dict itself (since the override dict in the o-env is a reference, not a copy, you can do this).  The change isn't big, I think it looks like this:
```py
--- a/SCons/Environment.py
+++ b/SCons/Environment.py
@@ -2666,7 +2666,10 @@ class OverrideEnvironment(Base):
         # set in the override dict, so don't spend time checking for existance.
         if not key.isidentifier():
             raise UserError(f"Illegal construction variable {key!r}")
-        self.__dict__['overrides'][key] = value
+        if key in self.__dict__['overrides']:
+            self.__dict__['overrides'][key] = value
+        else:
+            self.__dict__['__subject'][key] = value
         if key in self.__dict__['__deleted']:
             # it's no longer "deleted" if we set it
             self.__dict__['__deleted'].remove(key)
```
Is that what you meant?

Reply:
well... that wouldn't fully do (2), as it would go to the parent in that case.

Reply:
Hrmm. was looking at __setattr__... oopsie..
Still, not sure I agree with the current implementation.